File: src/main/java/me/yoursole/ctf/Commands/Start.java
package me.yoursole.ctf.Commands;

import me.yoursole.ctf.DataFiles.GameData;
import me.yoursole.ctf.DataFiles.GameLoop;
import me.yoursole.ctf.DataFiles.Items.Flag;
import me.yoursole.ctf.DataFiles.Utils;
import org.bukkit.*;
import org.bukkit.attribute.Attribute;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import org.bukkit.scoreboard.*;

import java.io.File;
import java.util.Map;
import java.util.UUID;

public class Start implements CommandExecutor {
    @Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        if(GameData.it==null){
            if(sender.isOp()){
                //
                Utils.generateNewWorlds();
                StructureType type = GameData.structureTypes.stream().skip((int) (GameData.structureTypes.size() * Math.random())).findFirst().orElse(null);
                Location spawn = Utils.getStructure(GameData.world,type);
                GameData.gameSpawnPoint=spawn;
                spawn.setY(GameData.world.getHighestBlockYAt(spawn));

                StructureType netherType = GameData.structureTypesNether.stream().skip((int) (GameData.structureTypesNether.size() * Math.random())).findFirst().orElse(null);
                GameData.netherMainPoint = Utils.getStructure(GameData.world_nether,netherType);
                GameData.world_nether.getWorldBorder().setSize(100);
                GameData.world_nether.getWorldBorder().setCenter(GameData.netherMainPoint);

                GameData.world.getWorldBorder().setCenter(spawn);
                GameData.world.getWorldBorder().setSize(200);

                for(Player player : Bukkit.getOnlinePlayers()){
                    Utils.sendPlayer(player,GameData.world);
                }

                //on player death, tp player to world spawn if they are playing the game
                //on player portal, tp to appropriate dimension
                //
                GameData.scores.clear();
                ScoreboardManager sb = Bukkit.getScoreboardManager();
                Scoreboard s = sb.getMainScoreboard();
                Objective o = s.getObjective(DisplaySlot.SIDEBAR);
                if (o != null) o.unregister();

                Player player = Bukkit.getOnlinePlayers().stream().skip((int) (Bukkit.getOnlinePlayers().size() * Math.random())).findFirst().orElse(null);
                player.getInventory().addItem(Flag.flag);
                player.setGlowing(true);
                GameData.itLoc=player.getLocation();
                GameData.it=player;
                player.sendMessage(ChatColor.GREEN+"You got the flag to start!");

                for(Player playera : Bukkit.getOnlinePlayers()){
                    GameData.scores.put(playera.getUniqueId(), 0);
                    if(!player.getDisplayName().equals(playera.getDisplayName())){
                        playera.sendMessage(ChatColor.GREEN + player.getDisplayName()+ " has received the flag");
                        playera.setGlowing(false);
                    }
                    playera.sendMessage(ChatColor.AQUA+"The selected structure is: "+ (type != null ? type.getName() : null) +"!");
                    playera.setHealth(playera.getAttribute(Attribute.GENERIC_MAX_HEALTH).getValue());
                    playera.getActivePotionEffects().clear();

                    //


                    //
                }
            }else{
                sender.sendMessage("don't u dare ._.");
            }
        }else{
            sender.sendMessage(ChatColor.RED+"The game is already running");
        }

        return true;

    }

}

File: src/main/kotlin/me/yoursole/ctf/commands/Start.kt
package me.yoursole.ctf.commands

import com.mojang.brigadier.CommandDispatcher
import me.yoursole.ctf.CTF
import me.yoursole.ctf.datafiles.FlagDropper
import me.yoursole.ctf.datafiles.GameData
import me.yoursole.ctf.datafiles.Utils.findSafeSpawnOrMake
import me.yoursole.ctf.datafiles.WorldManager
import me.yoursole.ctf.datafiles.WorldManager.getStructureNearSpawn
import me.yoursole.ctf.datafiles.WorldManager.sendToWorld
import net.minecraft.commands.CommandSourceStack
import net.minecraft.commands.Commands.*
import org.bukkit.Bukkit
import org.bukkit.attribute.Attribute
import org.bukkit.scoreboard.DisplaySlot
import kotlin.random.Random

object Start : BrigadierCommand {
    var id: Int = -1

    override fun register(dispatcher: CommandDispatcher<CommandSourceStack>) {
        dispatcher.register(literal("start").requiresOp().executes { ctx ->
            startGame(ctx.source, false)
        }.then(literal("startnow").executes { ctx ->
            startGame(ctx.source, true)
        }))
    }

    private fun startGame(source: CommandSourceStack, startImmediately: Boolean): Int {
        val sender = source.bukkitSender
        if (GameData.gameRunning) {
            sender.sendMessage("§cGame is already running!")
            return 0
        }
        WorldManager.generateNewWorlds()
        val type = GameData.structureTypes.random()
        val spawn = GameData.world.getStructureNearSpawn(type)!!
            .apply { y = GameData.world!!.getHighestBlockYAt(this).toDouble() }.findSafeSpawnOrMake()
        GameData.gameSpawnPoint = spawn
        val netherSpawn = GameData.world_nether.getStructureNearSpawn(GameData.structureTypesNether.random())!!
            .findSafeSpawnOrMake()
        GameData.netherMainPoint = netherSpawn
        GameData.world_nether!!.worldBorder.size = 100.0
        GameData.world_nether!!.worldBorder.center = GameData.netherMainPoint
        GameData.world!!.worldBorder.center = spawn
        GameData.world!!.worldBorder.size = 200.0
        GameData.scores.clear()
        Bukkit.getScoreboardManager().mainScoreboard.getObjective(DisplaySlot.SIDEBAR)?.unregister()
        for (player in Bukkit.getOnlinePlayers()) {
            player.sendToWorld(GameData.world)
            GameData.scores[player.uniqueId] = 0
            player.isGlowing = false
            player.sendMessage(
                "§bThe selected structure is: ${
                    type?.name?.split('_')?.joinToString(" ") { str -> str.replaceFirstChar { it.titlecaseChar() } }
                }!"
            )
            player.health = player.getAttribute(Attribute.GENERIC_MAX_HEALTH)!!.value
            player.activePotionEffects.clear()
            player.inventory.clear()
        }
        //set the timer
        GameData.timerMs = System.currentTimeMillis() + 15 * 60 * 1000 //15 minutes

        GameData.gameRunning = true
        id = Bukkit.getScheduler().scheduleSyncDelayedTask(
            CTF.instance,
            {
                FlagDropper.dropFlag(GameData.world!!,
                    GameData.gameSpawnPoint.blockX + Random.nextInt(-100, 100),
                    GameData.gameSpawnPoint.blockZ + Random.nextInt(-100, 100),
                    75)
            }, if (startImmediately) 1L else 15 * 60 * 20 //15 minutes
        )

        //on player death, tp player to world spawn if they are playing the game
        //on player portal, tp to appropriate dimension
        //
        return 1
    }
}